src/nfa/operations.cc

```diff
@@ -56,6 +56,126 @@ namespace {
         return lts_for_simulation.compute_simulation();
     }
 
+    bool vec_contains(std::vector<std::pair<State, State>> vec, std::pair<State, State> par){
+        for (size_t i = 0; i < vec.size(); i++){
+            if (vec[i] == par){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    Nfa reduce_testing(const Nfa& aut){
+        // ! Preprocessing
+        Nfa result_nfa;
+        Nfa reverted_nfa;
+
+        std::vector<std::pair<State, State>> result_sim {}; // R
+        std::vector<std::pair<State, State>> worklist {}; // Worklist
+
+        // Alphabet extraction
+        mata::OnTheFlyAlphabet alph;
+        aut.fill_alphabet(alph);
+        std::vector<Symbol> alph_syms = alph.get_alphabet_symbols().to_vector();
+
+        size_t no_states = aut.num_of_states();
+        size_t ***matrix;
+
+        // Matrix allocation
+        matrix = static_cast<size_t ***>(malloc(sizeof(size_t **) * alph_syms.size()));
+        for (size_t i = 0; i < alph_syms.size(); i++) {
+            matrix[i] = static_cast<size_t **>(malloc(sizeof(size_t *) * no_states));
+            for (size_t j = 0; j < no_states; j++) {
+                matrix[i][j] = static_cast<size_t *>(malloc(sizeof(size_t) * no_states));
+            }
+        }
+
+        reverted_nfa = revert(aut); // Reverted NFA
+        // ! End of preprocessing
+
+        // ! Initial refinement
+        for (size_t x = 0; x < alph_syms.size(); x++) {
+            for (size_t p = 0; p < no_states; p++) {
+                for (size_t q = 0; q < no_states; q++) {
+                    size_t q_size;
+                    size_t p_size;
+
+                    auto symbol_q = aut.delta[q].find(alph_syms[x]);
+                    if (symbol_q == aut.delta[q].end()) {
+                        matrix[x][p][q] = 0;
+                        q_size = 0;
+                    } else {
+                        q_size = (*symbol_q).num_of_targets();
+                        matrix[x][p][q] = q_size;
+                    }
+
+                    auto symbol_p = aut.delta[p].find(alph_syms[x]);
+                    if (symbol_q == aut.delta[p].end()) {
+                        p_size = 0;
+                    } else {
+                        p_size = (*symbol_p).num_of_targets();
+                    }
+                    if ((p_size != 0 && q_size == 0) || (aut.final.contains(p) && !aut.final.contains(q))) {
+                        if (!vec_contains(result_sim, std::pair(p,q))) {
+                            result_sim.push_back(std::pair(p,q)); // R append 
+                            worklist.push_back(std::pair(p,q)); // worklist append
+                        }
+                    }
+                }
+            }
+        }
+        // ! End of initional refinement
+
+        // ! Propagate until fixpoint
+        size_t worklist_size;
+        std::pair<State, State> working_pair;
+        while ((worklist_size = worklist.size()) != 0) {
```

Comment:
maybe just use `worklist.size()`? compiler should be able to optimize the two calls...

Reply:
Testing for (non-)emptiness should be done using `worklist.empty()`, optionally with `!` in front of the expression for non-emptiness. And we can then use `size()` inside the loop body.

Reply:
Testing for (non-)emptiness should be done using `worklist.empty()`, optionally with `!` in front of the expression for non-emptiness. And we can then use `size()` inside the loop body.